contracts/IntervalTokenStore.sol

```diff
@@ -0,0 +1,142 @@
+// IntervalTokenStore stores tokens for applications, which have discrete states increasing with time
+pragma solidity ^0.5.0;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "openzeppelin-solidity/contracts/math/Math.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+contract IntervalTokenStore {
+    using SafeMath for uint;
+
+    event Deposit(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event WithdrawRequest(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event Withdraw(
+        address user,
+        address token,
+        uint amount
+    );
+
+    // User => Token => BalanceState
+    mapping(address => mapping(address => BalanceState)) private balanceStates;
+
+    struct BalanceState {
+        uint256 balance;
+        PendingFlux pendingDeposits;
+        PendingFlux pendingWithdraws;
+    }
+
+    struct PendingFlux {
+        uint256 amount;
+        uint256 stateIndex;  // deposits will be processed for any currentStateIndex > stateIndex
+                              // withdraws are allowed for any currentStateIndex > stateIndex
+    }
+
+    uint256 public currentStateIndex=0;
+
+    function updateAndGetBalance(address user, address token) public returns(uint256) {
+        updateDepositsBalance(user, token);
+        uint balance = balanceStates[user][token].balance;
+        if (balanceStates[user][token].pendingWithdraws.stateIndex < currentStateIndex) {
+            balance -= Math.min(balanceStates[user][token].pendingWithdraws.amount, balance);
+        }
+        return balance;
+    }
+
+    function deposit(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token);
+        require(
+            ERC20(token).transferFrom(msg.sender, address(this), amount),
+            "Tokentransfer for deposit was not successful"
+        );
+        balanceStates[msg.sender][token].pendingDeposits.amount += amount;
+        balanceStates[msg.sender][token].pendingDeposits.stateIndex = currentStateIndex;
+        emit Deposit(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdrawRequest(address token, uint amount) public {
+        balanceStates[msg.sender][token].pendingWithdraws = PendingFlux({ amount: amount, stateIndex: currentStateIndex });
+        emit WithdrawRequest(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdraw(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token);
```

Comment:
Maybe comment that this is necessary in order to avoid negative balance (deposits might have to be credited in order to withdraw).